### PR TITLE
#2 - Fix multiple open report call when report screen is mounted

### PR DESCRIPTION
**Pull request by bernhardoj**
### Details

We call OpenReport in both ReportScreen and ReportActionsView. This PR fix it.

### Fixed Issues

$ https://github.com/Expensify/App/issues/41307
$ https://github.com/Expensify/App/issues/39673
PROPOSAL: https://github.com/Expensify/App/issues/39673#issuecomment-2041127917

### Tests

Same as QA Steps
- [x] Verify that no errors appear in the JS console

### Offline tests

Same as QA Steps
### QA Steps

Web/Desktop
1. Open developer tools
2. Switch to the Network tab
3. Sign out and sign in
4. Wait until the app load completes
5. In the network tab, verify there is only 1 OpenReport call and the report chat is shown

Android/iOS/mWeb
1. Open developer tools (not available in Android/iOS)
2. Switch to the Network tab
3. Open any report from LHN
4. In the network tab, verify there is only 1 OpenReport call and the report chat is shown

Next test for all platforms
1. Open a thread chat
2. Press the "From ..." text link
3. Verify the report screen won't show a blank screen briefly
4. Open an invalid reportID (such as 123) that you never load before in the device
5. Verify the skeleton shows and not found view will show after loading completes
6. Open a valid reportID that you don't have access to
7. Verify the not found view will show after loading completes
- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/50919443/cc291c32-c14f-4165-824a-d42424321d21

https://github.com/Expensify/App/assets/50919443/4d57c46d-6178-4147-8854-b640da084617

https://github.com/Expensify/App/assets/50919443/4a90d3d5-7e66-44fb-8c78-f29c30ffc3bd

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/50919443/32945c19-b93a-4346-9f76-89734a757f17

https://github.com/Expensify/App/assets/50919443/dc09f88e-51f7-485c-9ac7-49ed24172353

https://github.com/Expensify/App/assets/50919443/1fb3e437-8749-4e30-8a68-779088dac708

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/50919443/9479ef23-217f-4540-9c10-8bc5b6243050

https://github.com/Expensify/App/assets/50919443/b955f55a-3d41-4c90-b2d0-261c814b379d

https://github.com/Expensify/App/assets/50919443/9ed51251-f632-4187-9b1f-969491dcc4b6

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/50919443/9581f393-9d3a-47c0-9247-3f4c17394de0

https://github.com/Expensify/App/assets/50919443/bf79e727-3334-4d56-b7ed-5f9a31db8c24

https://github.com/Expensify/App/assets/50919443/c0c7592b-eb04-4ab6-a7ae-b5a524de0d90

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/50919443/ba52c305-7cf5-403d-b20c-8cd9ef361b79

https://github.com/Expensify/App/assets/50919443/0c8484ba-1447-4941-b6c4-b70d284f328a

https://github.com/Expensify/App/assets/50919443/b3e136e5-84d4-4dc6-84ea-447ddcba7f7d

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/50919443/802ed9fc-7c38-4209-8680-ce60dab0b56e

https://github.com/Expensify/App/assets/50919443/6eeaeee8-2236-4232-a7f3-27d895829a3d

https://github.com/Expensify/App/assets/50919443/9878c011-650d-469e-a2a6-a6b4b8caa650

</details>

### CODE REVIEW

**Comment by bernhardoj** [comment]
The reason https://github.com/Expensify/App/issues/41307 happens is because if the report is loading, we return an empty report actions. Idk why we decide to do that.

https://github.com/Expensify/App/blob/8661f6a8234fd2a6c6b99dce726593b41ec7cc68/src/pages/home/report/ReportActionsView.tsx#L112
https://github.com/Expensify/App/blob/8661f6a8234fd2a6c6b99dce726593b41ec7cc68/src/pages/home/report/ReportActionsView.tsx#L165-L171

That's why we don't call OpenReport inside ReportActionsView if we are linking to a message.
https://github.com/Expensify/App/blob/8661f6a8234fd2a6c6b99dce726593b41ec7cc68/src/pages/home/report/ReportActionsView.tsx#L240-L247

So, I decided to do the same.

File: src/pages/home/ReportScreen.tsx [hunk 1]
```diff
@@ -465,7 +458,11 @@ function ReportScreen({
         Timing.end(CONST.TIMING.CHAT_RENDER);
         Performance.markEnd(CONST.TIMING.CHAT_RENDER);
 
-        fetchReportIfNeeded();
+        // Call OpenReport only if we are not linking to a message or the report is not available yet
+        if (!reportActionIDFromRoute || !report.reportID) {
+            fetchReportIfNeeded();
+        }
```
[the remark this answers is not in the document]
Comment (neil-marcellini) [reply src/pages/home/ReportScreen.tsx:464]: If we are linking to a message why don't we need to call OpenReport? Can we put this logic inside fetchReportIfNeeded instead?

[the remark this answers is not in the document]
Comment (neil-marcellini) [reply src/pages/home/ReportScreen.tsx:464]: Ah I see we have a separate effect that calls OpenReport whenever `reportActionIDFromRoute` changes. The comment should explain that.

File: src/pages/home/report/ReportActionsView.tsx [hunk 2]
```diff
@@ -237,24 +237,6 @@ function ReportActionsView({
     const oldestReportAction = useMemo(() => reportActions?.at(-1), [reportActions]);
     const hasCreatedAction = oldestReportAction?.actionName === CONST.REPORT.ACTIONS.TYPE.CREATED;
 
-    useEffect(() => {
-        if (reportActionID) {
-            return;
-        }
-
-        const interactionTask = InteractionManager.runAfterInteractions(() => {
```
Comment (neil-marcellini) [inline src/pages/home/report/ReportActionsView.tsx:245]: What is this interactionTask for and is it ok that we won't have it in ReportScreen?

**Review by neil-marcellini — CHANGES_REQUESTED** [review]
👋 Hi, I'm jumping in to review since this seems valuable. The logic for this area is pretty involved and I don't claim to fully understand it, so I'm asking some questions.

Let's pleas also add a lot more tests, such as opening a link to a message, opening an invalid reportID, opening a valid report link, opening a non-existent report link.

File: src/pages/home/report/ReportActionsView.tsx (hunk 2 above)
Comment (bernhardoj) [reply src/pages/home/report/ReportActionsView.tsx:245]: I found that it's added from this [performance improvement PR](https://github.com/Expensify/App/pull/38255). Should we put it inside the interaction task?
https://github.com/Expensify/App/blob/1c651a1ac7c6af25df9459861ba4128e8d50e2fb/src/pages/home/ReportScreen.tsx#L468-L471

**Comment by bernhardoj** [comment]
Added test for the invalid/non-existing report

File: src/pages/home/ReportScreen.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (neil-marcellini) [reply src/pages/home/ReportScreen.tsx:464]: There's also another one here, hmm.

https://github.com/Expensify/App/blob/7e658961b2300bf1ae75197d3d988a0e56d1d16b/src/pages/home/ReportScreen.tsx#L401-L403

File: src/pages/home/report/ReportActionsView.tsx (hunk 2 above)
Comment (neil-marcellini) [reply src/pages/home/report/ReportActionsView.tsx:245]: Please do a test with it in the interaction task and without, and show the results with videos and or timing logs.

**Review by neil-marcellini — COMMENTED** [review]
Thanks for your replies. There's still a lot I'm unsure of, so I suggest some more tests to verify the behavior for both yourself and especially us reviewers.

Also please add steps distinguishing between opening an invalid reportID (such as `0`), opening a valid report link, and opening a non-existent report link (some random valid id that doesn't exist). "Open an invalid chat" is a bit vague.

File: src/pages/home/ReportScreen.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (bernhardoj) [reply src/pages/home/ReportScreen.tsx:464]: That's just a function.

File: src/pages/home/report/ReportActionsView.tsx (hunk 2 above)
Comment (bernhardoj) [reply src/pages/home/report/ReportActionsView.tsx:245]: Updated to put it inside interaction task.

Without interaction task:

https://github.com/Expensify/App/assets/50919443/37915ee0-b09a-4f1d-a7e5-38c6bde741f3

With interaction task:

https://github.com/Expensify/App/assets/50919443/fc492e63-7bdc-43fc-a305-d087763f4396


Here is the log result of `switch_report` timing.
https://github.com/Expensify/App/blob/ee8f9c661f269f405a5ae8ba47ea07f42a306c6c/src/pages/home/report/ReportActionsView.tsx#L385-L401


| Without interaction task  | With interaction task |
| ------------- | ------------- |
| 1033  | 1025  |
| 879  | 994 |
| 1152 | 1078 |
| 1692 | 1652 |
| Avg: 1189 | Avg: 1187.25 |

Comment (neil-marcellini) [reply src/pages/home/report/ReportActionsView.tsx:245]: Interesting thanks so much! The timing results don't look too different, but the video shows it's better with the interaction task.

File: src/pages/home/ReportScreen.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (neil-marcellini) [reply src/pages/home/ReportScreen.tsx]: 🤦‍♂️ oh right

**Review by neil-marcellini — DISMISSED** [review]
Ok, I _think_ it should be good now. I still don't see a test for reportID 0. Leaving it to C+ and the managing engineer for final approvals and merge.

**Comment by abdulrahuman5196** [comment]
@bernhardoj The skeleton view is flickering when moving to a report after sign-in. Could you kindly check on this?


https://github.com/Expensify/App/assets/46707890/ac06a0f3-915c-43e8-9d5d-278111d2b3af

**Comment by abdulrahuman5196** [comment]
@neil-marcellini Could you kindly check on the above?

**Comment by bernhardoj** [comment]
@abdulrahuman5196 updated

**Comment by abdulrahuman5196** [comment]
Checking now

**Comment by abdulrahuman5196** [comment]
@bernhardoj I am seeing multiple openReport calls after signin.


https://github.com/Expensify/App/assets/46707890/ebe4120c-5be4-4688-96aa-a9d6e1a4444e

**Comment by bernhardoj** [comment]
That's really weird, i can't reproduce that at all.

https://github.com/Expensify/App/assets/50919443/d09104a9-5db5-40d0-b780-93c4c9985dda

**Comment by neil-marcellini** [comment]
@abdulrahuman5196 would you please re-test and provide specific test steps if you still experience the bug?

**Comment by abdulrahuman5196** [comment]
Checking now again

**Comment by abdulrahuman5196** [comment]
After latest merge I don't see this issue - https://github.com/Expensify/App/pull/41427#issuecomment-2107292181

Finishing checklist.

**Comment by abdulrahuman5196** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>






https://github.com/Expensify/App/assets/46707890/d69029be-1c68-4792-aea4-13ca322cad84


</details>

<details>
<summary>Android: mWeb Chrome</summary>





https://github.com/Expensify/App/assets/46707890/14cde376-6965-4d35-b4f0-d72ddbb4c5f8




</details>

<details>
<summary>iOS: Native</summary>




https://github.com/Expensify/App/assets/46707890/52066e0f-7eea-49f7-9981-ced81d8978bb




</details>

<details>
<summary>iOS: mWeb Safari</summary>





https://github.com/Expensify/App/assets/46707890/8aa89ceb-33b4-4cd0-856c-6a1e32ce4790




</details>

<details>
<summary>MacOS: Chrome / Safari</summary>





https://github.com/Expensify/App/assets/46707890/44add320-5df2-4483-ac68-36122f4fb24d


</details>

<details>
<summary>MacOS: Desktop</summary>





https://github.com/Expensify/App/assets/46707890/10c3511b-6b31-4be5-a6d4-0c511304d257



</details>

**Review by abdulrahuman5196 — APPROVED** [review]
Changes looks good and works well. Reviewers checklist is also complete.

All yours. @neil-marcellini / @flodnv 

🎀 👀 🎀
C+ Reviewed

**Review by neil-marcellini — APPROVED** [review]

**Review by flodnv — APPROVED** [review]

**Pull request merged**